### DISCUSSION TITLE
WebAssembly & Connections from Client to Outside World?

### DISCUSSION THREAD

**Discussion by kuiperzone** [AvaloniaUI/Avalonia#14074] [Q&A]
In the not too distant future, I will be interested in building Avalonia apps for WebAssembly and have them run in a client browser. Critical to this will be the ability for the app to connect to the outside world. Specifically, I'm thinking about my MySQL or SSH.NET connections to a server somewhere.

What's the deal with WebAssembly and this type of connectivity?

Is this going to be possible?

If it's not possible now, will it be possible in the future? I read that WebSockets are possible with WebAssembly?

Cheers

**Comment by thevortexcloud** [comment]
Typically, you would not give the client of a web app direct database access. That is a bad idea for several reasons. Instead you make some other server/API the client talks to, which then talks to the database/other service on behalf of the client. In which case you could probably get away with a simple ASP Core site and REST API in most cases. `HttpClient` should work fine as far as I know.

**Reply by kuiperzone** [reply]
Thank you. I agree generally with what you say about client database connectivity. But assume for a moment that I was building something like MySQL Workbench running over a local network, which wouldn't be too far off.

And my question still stands...

Is it going to be possible to establish direct outward socket connections from a client WebAssembly?

Is `ClientWebSocket` possible?

Or is `HttpClient` all I would have from a WebAssembly?

**Reply by thevortexcloud** [reply]
I am no web assembly expert, but I believe Web Sockets in general work, as that's what Blazor uses for most of it's stuff. But I have never tried it myself. The other problem you would run into is a lot of database libraries (excluding raw ADO) rely heavily on reflection/dynamic code and are fairly broken when using AOT and trimming. Most of them are also just wrappers around a native C library. Which means they would need to be compiled to be web assembly compatible (most are not designed to do that). Microsoft has fairly brief overview here of what is and is not supported:

https://github.com/dotnet/runtime/blob/main/src/mono/wasm/features.md

So I _think_ no matter what, you would need to proxy a database connection unless you can find a compatible driver. The only one I know of the top of my head though is sqlite, which was ported to web assembly. I am not sure if it works with .NET though.

**Reply by kuiperzone** [reply]
Thank you. That's really helpful.

**Comment by kuiperzone** [comment]
Just a follow on thought...

I was thinking that gRPC uses HTTP/2 and so it might be possible to use gRPC with WebAssembly. Looks like it is:

https://azure.github.io/AppService/2021/03/15/How-to-use-gRPC-Web-with-Blazor-WebAssembly-on-App-Service.html

Doesn't give generic sockets etc., but it might be relevant to building API proxies.

Am open to comments...

**Reply by thevortexcloud** [reply]
From memory, gRPC does not work very well for handling large amounts of data. Which could be a problem if you are trying to return a very large query result (which is likely to happen if someone is storing binary data inside their database). But it has been a long time since I used it so I could be misremembering.

**Comment by satial-interfaces** [comment]
I use a gRPC to ADO.NET (database) bridge project at:

https://github.com/james-brighton/GrpcDataProvider

It includes an Avalonia WebAssembly example.
